src/site/content/en/blog/more-capable-form-controls/index.md

```diff
@@ -0,0 +1,305 @@
+
+```js
+this._internals.setFormValue(this._value, 
+    this._mode + '/' + this._value);
+```
+
+Your code would need to restore its state based on the stored state value.
+
+```js
+formStateRestoreCallback(state, mode) {
+  if (mode == 'restore') {
+    // expects a state parameter in the form 'controlMode/value'
+    [controlMode, value] = state.split('/');
+    this._mode = controlMode;
+    this._value = value;
+  }
+  // Chrome currently doesn't handle autofill for form-associated 
+  // custom elements. In the autofill case, you might need to handle
+  // a raw value.
+}
+```
+
+In the case of a simpler control (for example a number input), the value is probably sufficient to restore the control to its previous state. If you omit `state` when calling `setFormValue`, then the value is passed to `formStateRestoreCallback`. 
+
+```js
+formStateRestoreCallback(state, mode) {
+  // Simple case, restore the saved value
+  this._value = state;
+}
+```
+
+### A working example
+
+The following example puts together many of the features of form-associated custom elements.
+Be sure to run it on Chrome 77+ (Canary or Beta) to see the API in action.
+
+<div class="glitch-embed-wrap" style="height: 420px; width: 100%;">
+  <iframe
+    allow="geolocation; microphone; camera; midi; encrypted-media"
+    src="https://glitch.com/embed/#!/embed/form-associated-ce?path=my-control.js&amp;previewSize=0"
+    alt="formdata event demo on Glitch"
+    style="height: 100%; width: 100%; border: 0;">
+  </iframe>
+</div>
+
+## Feature detection 
+
+You can use feature detection to determine whether the `formdata` event and form-associated custom elements are available. There are currently no polyfills released for either feature. In both cases, you can fall back to adding a hidden form element to propagate the control's value to the form. Many of the more advanced features of form-associated custom elements will likely be difficult or impossible to polyfill.
+
+```js
+if ('FormDataEvent' in window) {
+  // formdata event is supported
+}
+
+if ('ElementInternals' in window && 
+    'setFormData' in window.ElementInternals) {
+  // Form-associated custom elements are supported
+}
+```
+
+## Conclusion
+
+The `formdata` event and form-associated custom elements provide new tools for creating custom form controls. 
+
+The `formdata` event doesn't give you any new capabilities, but it gives you an interface for adding your form data to the submit process, without having to create a hidden `<input>` element.
+
+The form-associated custom elements API provides a new set of capabilities for making custom form controls that work like built-in form controls.
+
+_Hero image by Oudom Pravat on Unsplash._
```

Comment:
Just curious if you're able to link this back to the original image? In the future I may use these when we 2x all the headers.

Reply:
Sure. I hadn't seen anyone else doing that, so I didn't include a link... (In fact, I had to search long and hard to find anyone crediting their hero image.)

For the sake of consistency, how do you want to link it up?

_Hero image by Oudom Pravat on [Unsplash](https://unsplash.com/photos/yQi4mAFmRew)._

Or

_[Hero image](https://unsplash.com/photos/yQi4mAFmRew) by Oudom Pravat on Unsplash._

Or 

[_Hero image by Oudom Pravat on Unsplash._](https://unsplash.com/photos/yQi4mAFmRew)

Reply:
I went with #1 here, can update if desired.